### PR TITLE
Update shardLeaders cache to concurrent map type

### PR DESCRIPTION
**Pull request by bigsheeper**
Signed-off-by: bigsheeper <yihao.dai@zilliz.com>

issues: https://github.com/milvus-io/milvus/issues/19454

### CODE REVIEW

File: internal/proxy/task_policies.go [hunk 1]
```diff
@@ -14,35 +14,36 @@ import (
...
-		if len(leaders) <= 1 {
+func updateShardsWithRoundRobin(shardsLeaders ShardLeaders) {
+	for items := range shardsLeaders.IterBuffered() {
+		if len(items.Val) <= 1 {
 			continue
 		}
-
-		shardsLeaders[channelID] = append(leaders[1:], leaders[0])
+		items.Val = append(items.Val[1:], items.Val[0])
+		shardsLeaders.Set(items.Key, items.Val)
```
Comment (congqixia) [inline internal/proxy/task_policies.go:30]: If we need COW here, we shall clone the values of items and set it back instead of writing it directly.

Comment (bigsheeper) [reply internal/proxy/task_policies.go]: solved

**Review by congqixia — APPROVED** [review]
/lgtm

**Pull request closed without merging**